File: app/src/main/java/com/niro/niroapp/network/ApiClient.java
package com.niro.niroapp.network;

import java.util.concurrent.TimeUnit;

import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class ApiClient {

    private static Retrofit retrofit = null;

    // Get client object
    public static Retrofit getClient() {

        if (retrofit == null) {
            retrofit = new Retrofit.Builder()
                    .baseUrl("https://niroapp.df.r.appspot.com")
                    .addConverterFactory(GsonConverterFactory.create())
                    .client(getHttpClient())
                    .build();
        }
        return retrofit;
    }

    // Set client
    public static OkHttpClient getHttpClient() {


        OkHttpClient httpClient = new OkHttpClient.Builder()
                .connectTimeout(60, TimeUnit.SECONDS)
                .writeTimeout(60, TimeUnit.SECONDS)
                .readTimeout(60, TimeUnit.SECONDS)
                .build();

        return httpClient;
    }

    public static ApiInterface getAPIInterface() {
        return getClient().create(ApiInterface.class);
    }

}

File: app/src/main/java/com/niro/niroapp/network/ApiClient.kt
package com.niro.niroapp.network

import android.content.Context
import com.niro.niroapp.utils.NiroAppUtils.getToken
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.util.concurrent.TimeUnit

object ApiClient {

    private var retrofit: Retrofit? = null

    // Get client object
    private fun getClient(context: Context?): Retrofit? {
        if (retrofit == null) {
            retrofit = Retrofit.Builder()
                .baseUrl("https://niroapp.df.r.appspot.com")
                .addConverterFactory(GsonConverterFactory.create())
                .client(getHttpClient(context))
                .build()
        }
        return retrofit
    }

    // Set client
    private fun getHttpClient(context: Context?): OkHttpClient {
        return OkHttpClient.Builder()
            .connectTimeout(60, TimeUnit.SECONDS)
            .writeTimeout(60, TimeUnit.SECONDS)
            .readTimeout(60, TimeUnit.SECONDS)
            .addInterceptor { chain ->
                val newRequest = chain.request().newBuilder()
                    .addHeader("Authorization", getToken(context!!))
                    .build()
                chain.proceed(newRequest)
            }
            .build()
    }


    fun getAPIInterface(context: Context?): ApiInterface? {
        return getClient(context)?.create(ApiInterface::class.java)
    }
}